### PR TITLE
#58 Generics commit

### PR DESCRIPTION
**Pull request by ivelin980**

### CODE REVIEW

File: javaAdvanced_generics/src/customlist/CustomList.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+	@SuppressWarnings("unchecked")
+	public E getMax() {
+		return Arrays.stream(this.elements).limit(this.size).map(e -> (E) e).max(Comparator.comparing(e -> e))
+				.orElse(null);
+	}
+
+	@SuppressWarnings("unchecked")
+	public E getMin() {
+		return Arrays.stream(this.elements).limit(this.size).map(e -> (E) e).min(Comparator.comparing(e -> e))
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_generics/src/customlist/CustomList.java:98]: It was part of the course lecture.

File: javaAdvanced_generics/src/customlistsorter/CustomList.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	public void sort() {
+		new Sorter<E>().sort(this);
+	}
+
+	@Override
+	public String toString() {
+		for (int i = 0; i < this.size; i++) {
+			System.out.println(this.elements[i]);
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_generics/src/customlistsorter/CustomList.java:109]: fixed

File: javaAdvanced_generics/src/customlist/CustomList.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+				.orElse(null);
+	}
+
+	@Override
+	public String toString() {
+		for (int i = 0; i < this.size; i++) {
+			System.out.println(this.elements[i]);
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_generics/src/customlist/CustomList.java:105]: fixed

File: javaAdvanced_generics/src/customlistsorter/CustomList.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+			throw new IndexOutOfBoundsException();
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	private E getAt(int index) {
+		return (E) this.elements[index];
+	}
+
+	E remove(int index) {
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_generics/src/customlistsorter/CustomList.java:46]: fixed

File: javaAdvanced_generics/src/customlistsorter/CustomList.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+		for (int i = 0; i < this.size; i++) {
+			E at = this.getAt(i);
+			if (at.equals(obj)) {
+				return i;
+			}
+		}
+		return -1;
+	}
+
+	void swap(int firstIndex, int secondIndex) {
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_generics/src/customlistsorter/CustomList.java:72]: fixed

File: javaAdvanced_generics/src/customlistsorter/Sorter.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package customlistsorter;
+
+public class Sorter<T extends Comparable<T>> {
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_generics/src/customlistsorter/Sorter.java:3]: fixed

File: javaAdvanced_generics/src/genericscale/Scale.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+	
+	public T getHeavier() {
+		if(left.compareTo(right)>0) {
+			return left;
+		}
+		else if(left.compareTo(right)<0) {
+			return right;
+		}
+		else {
+		return null;
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_generics/src/genericscale/Scale.java:20]: fixed

File: javaAdvanced_generics/src/listutilities/ListUtils.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+import java.util.Collections;
+import java.util.List;
+
+public class ListUtils {
+	private ListUtils() {
+	}
+
+	public static <T extends Comparable<T>> T getMin(List<T> list) {
+		if (list.isEmpty()) {
+			throw new IllegalArgumentException();
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_generics/src/listutilities/ListUtils.java:12]: fixed

**Review by ngavrailov — APPROVED** [review]

File: javaAdvanced_generics/src/customlist/CustomList.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	public E get(int index) {
+		ensureIndex(index);
+		return getAt(index);
+	}
+
+	private void ensureIndex(int index) {
+		if (index < 0 || index >= this.size) {
+			throw new IndexOutOfBoundsException();
```
[the remark this answers is not in the document]
Comment (ivelin980) [reply javaAdvanced_generics/src/customlist/CustomList.java:37]: fixed

**Pull request merged**